assets/training/finetune_acft_hf_nlp/environments/acpt/context/finetune_run.py

```diff
@@ -367,32 +369,117 @@ def add_task_specific_params(cmd: List[str], task_name: str, component_name: str
                     add_optional_input(cmd, param)
 
 
-def _run_subprocess_cmd(cmd: List[str], component_name: str, completion_files_folder: str):
+def is_main_process():
+    """
+    To check if current process running this is master or rank_0 process.
+
+    :return: Boolean for whether this process is master.
+    """
+    return os.environ.get('AZUREML_PROCESS_NAME', 'main') in {'main', 'rank_0'}
+
+
+def wait_at_barrier(barrier_file, num_processes):
+    """
+    Control will halt till number to process reaching the execution point is less than a given number.
+
+    barrier_file: File used to create execution barrier.
+    num_processes: Number of process which need to reach barrier point.
+    """
+    process_name = os.environ.get('AZUREML_PROCESS_NAME', 'main')
+    with open(barrier_file, 'a+') as f:
+        portalocker.lock(f, portalocker.LOCK_EX)
+        try:
+            f.seek(0)
+            lines = f.readlines()
+            process_count = len(lines)
+            if process_count < num_processes:
+                f.write(f"{os.getpid()} reached the barrier\n")
+                f.flush()
+                portalocker.unlock(f)
+                logger.info(f'Process {process_name} at barrier, count is {process_count},')
+                logger.info(f'in barrier file {barrier_file}')
+        finally:
+            portalocker.unlock(f)
+        with open(barrier_file, 'a+') as f:
+            while len(lines) < num_processes:
+                portalocker.lock(f, portalocker.LOCK_EX)
+                f.seek(0)
+                lines = f.readlines()
+                portalocker.unlock(f)
+                time.sleep(0.5)  # Polling interval
+        logger.info(f"Process {os.getpid()} has passed barrier with name {process_name}")
+
+
+def _run_subprocess_cmd(cmd: List[str], component_name: str, completion_files_folder: str,
+                        single_run=True, number_of_processes=1):
     """Run the subprocess command."""
     logger.info(f"Starting the command: {cmd}")
     completion_file = Path(completion_files_folder, f"{component_name}.complete.txt")
+    barrier_file = Path(completion_files_folder, f"{component_name}.barrier.txt")
+    Path(completion_files_folder).mkdir(parents=True, exist_ok=True)

```

Comment:
if I understand correctly, a single completion file is shared among all the processes of this component? In that case if one of them has created the completion file, it is possible that all others will end up doing nothing due to the if statement below?

Reply:
> if I understand correctly, a single completion file is shared among all the processes of this component? In that case if one of them has created the completion file, it is possible that all others will end up doing nothing due to the if statement below?

yes, in this case , it is for creation of completion folder path. In case path already existing the Path library is hopefully not costly because of exist_ok argument ; If we don't use this way we need to either check if rank is 0 and get it created by it then if rank1 is scheduled first and it checks path it might get to error, we'll need to solve that scenario. Or we need an if condition to check for existence and create even in this case the present expression is locally of same cost.

Reply:
My question was specifically around completion file itself (and not the folder). It seems a single completion file is shared among all the workers. In that case it is possible that one of the workers succesfully completes its work, writes out the completion file before the 2nd worker has started. When the 2nd worker starts, it will see that the completion file exists and not do anything.

Reply:
> My question was specifically around completion file itself (and not the folder). It seems a single completion file is shared among all the workers. In that case it is possible that one of the workers succesfully completes its work, writes out the completion file before the 2nd worker has started. When the 2nd worker starts, it will see that the completion file exists and not do anything.

I didn't see such thing happening but this check seems to be present in older code